File: onebusaway-android/src/google/java/org/onebusaway/android/map/googlemapsv2/bike/BikeInfoWindow.java
/*
 * Copyright (C) 2017 Rodrigo Carvalho (carvalhorr@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.map.googlemapsv2.bike;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.Marker;

import org.onebusaway.android.R;
import org.opentripplanner.routing.bike_rental.BikeRentalStation;

/**
 * InfoWindow displayed when a floating bike or a bike station is clicked on the map.
 *
 * Created by carvalhorr on 6/16/17.
 */
public class BikeInfoWindow implements GoogleMap.InfoWindowAdapter {

    private View bikeInfoWindowView;

    private BikeStationsInfo bikeStationsInfo;

    public BikeInfoWindow(Context content, BikeStationsInfo bikeStationsInfo) {
        bikeInfoWindowView = LayoutInflater.from(content).inflate(R.layout.bike_info_window, null);
        this.bikeStationsInfo = bikeStationsInfo;
    }

    @Override
    public View getInfoWindow(Marker marker) {
        return null;
    }

    @Override
    public View getInfoContents(Marker marker) {
        BikeRentalStation station = bikeStationsInfo.getBikeStationOnMarker(marker);
        setBikeStationName(station.name);
        setNumberBikesAvailable(station.bikesAvailable);
        setNumberSpacesAvailable(station.spacesAvailable);
        return bikeInfoWindowView;

    }

    /**
     * Set the bike station name in the corresponding view.
     * @param name Name of the bike station
     */
    private void setBikeStationName(String name) {
        TextView stationName = (TextView) bikeInfoWindowView.findViewById(R.id.bikeStationName);
        stationName.setText(name);
    }

    /**
     * Set the  number of available bikes in the corresponding view
     * @param numberBikes number of bikes available
     */
    private void setNumberBikesAvailable(int numberBikes) {
        TextView bikesAvailable = (TextView) bikeInfoWindowView.findViewById(R.id.numberBikes);
        bikesAvailable.setText(String.valueOf(numberBikes));
    }

    /**
     * Set the number of spaces available to dropoff bikes
     * @param numberSpaces number of spaces available
     */
    private void setNumberSpacesAvailable(int numberSpaces) {
        TextView spacesAvailable = (TextView) bikeInfoWindowView.findViewById(R.id.numberRacks);
        spacesAvailable.setText(String.valueOf(numberSpaces));
    }

    /**
     * Interface to allow get the BikeRentalStation associated with a marker.
     */
    public interface BikeStationsInfo {
        BikeRentalStation getBikeStationOnMarker(Marker marker);
    }

}

File: onebusaway-android/src/main/java/org/onebusaway/android/map/compose/BikeInfoWindow.kt
/*
 * Copyright (C) 2026 Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.map.compose

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.onebusaway.android.R
import org.opentripplanner.routing.bike_rental.BikeRentalStation

// Info-window contents sit on a white bubble, so text uses fixed dark colors.
private val BikePrimary = Color(0xDE000000)
private val BikeSecondary = Color(0x99000000)

/**
 * The bike-station marker info-window content (shared across map flavors): a floating bike shows just
 * its name; a docking station shows its name plus available bikes and open spaces. On Google it is the
 * content of a `MarkerInfoWindowContent` (the SDK draws the white bubble); on MapLibre it is hosted in
 * a `ComposeView` wrapped in its own bubble.
 */
@Composable
fun BikeInfoWindow(station: BikeRentalStation) {
    Column(modifier = Modifier.padding(8.dp)) {
        Text(
            text = station.name,
            color = BikePrimary,
            fontWeight = FontWeight.Bold,
            fontSize = 14.sp,
        )
        if (station.isFloatingBike) {
            Text(
                text = stringResource(R.string.floating_bike_title),
                color = BikeSecondary,
                fontSize = 12.sp,
            )
        } else {
            Row(
                modifier = Modifier.padding(top = 2.dp),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                BikeCount(stringResource(R.string.bike_info_window_bikes_title), station.bikesAvailable)
                BikeCount(stringResource(R.string.bike_info_window_spaces_title), station.spacesAvailable)
            }
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = stringResource(R.string.bike_station_title),
                    color = BikeSecondary,
                    fontSize = 12.sp,
                )
                Icon(
                    painter = painterResource(R.drawable.ic_navigation_chevron_right),
                    contentDescription = null,
                    tint = BikeSecondary,
                    modifier = Modifier.size(20.dp),
                )
            }
        }
    }
}

@Composable
private fun BikeCount(title: String, count: Int) {
    Column {
        Text(text = title, color = BikeSecondary, fontSize = 11.sp)
        Text(text = count.toString(), color = BikePrimary, fontSize = 14.sp, fontWeight = FontWeight.Bold)
    }
}
